File: go/indicators/patrickmulloy/tripleexponentialmovingaverage/tripleexponentialmovingaverage.go
package tripleexponentialmovingaverage

//nolint: gofumpt
import (
	"fmt"
	"math"
	"sync"

	"zpano/entities"
	"zpano/indicators/core"
	"zpano/indicators/core/outputs"
)

// TripleExponentialMovingAverage computes the Triple Exponential Moving Average (TEMA),
// a smoothing indicator with less lag than a straight exponential moving average.
//
// The TEMA was developed by Patrick G. Mulloy and is described in two articles:
//
//	❶ Technical Analysis of Stocks & Commodities v.12:1 (11-19), Smoothing Data With Faster Moving Averages.
//	❷ Technical Analysis of Stocks & Commodities v.12:2 (72-80), Smoothing Data With Less Lag.
//
// The calculation is as follows:
//
//	EMA¹ᵢ = EMA(Pᵢ) = αPᵢ + (1-α)EMA¹ᵢ₋₁ = EMA¹ᵢ₋₁ + α(Pᵢ - EMA¹ᵢ₋₁), 0 < α ≤ 1
//	EMA²ᵢ = EMA(EMA¹ᵢ) = αEMA¹ᵢ + (1-α)EMA²ᵢ₋₁ = EMA²ᵢ₋₁ + α(EMA¹ᵢ - EMA²ᵢ₋₁), 0 < α ≤ 1
//	EMA³ᵢ = EMA(EMA²ᵢ) = αEMA²ᵢ + (1-α)EMA³ᵢ₋₁ = EMA³ᵢ₋₁ + α(EMA²ᵢ - EMA³ᵢ₋₁), 0 < α ≤ 1
//	TEMAᵢ = 3(EMA¹ᵢ - EMA²ᵢ) + EMA³ᵢ
//
// The very first EMA value (the seed for subsequent values) is calculated differently.
// This implementation allows for two algorithms for this seed.
//
//	❶ Use a simple average of the first 'period'. This is the most widely documented approach.
//	❷ Use first sample value as a seed. This is used in Metastock.
type TripleExponentialMovingAverage struct {
	mu sync.RWMutex
	core.LineIndicator
	smoothingFactor float64
	sum             float64
	ema1            float64
	ema2            float64
	ema3            float64
	length          int
	length2         int
	length3         int
	count           int
	firstIsAverage  bool
	primed          bool
}

// NewTripleExponentialMovingAverageLength returns an instnce of the indicator
// created using supplied parameters based on length.
func NewTripleExponentialMovingAverageLength(
	p *TripleExponentialMovingAverageLengthParams,
) (*TripleExponentialMovingAverage, error) {
	return newTripleExponentialMovingAverage(p.Length, math.NaN(), p.FirstIsAverage,
		p.BarComponent, p.QuoteComponent, p.TradeComponent)
}

// NewTripleExponentialMovingAverageSmoothingFactor returns an instnce of the indicator
// created using supplied parameters based on smoothing factor.
func NewTripleExponentialMovingAverageSmoothingFactor(
	p *TripleExponentialMovingAverageSmoothingFactorParams,
) (*TripleExponentialMovingAverage, error) {
	return newTripleExponentialMovingAverage(0, p.SmoothingFactor, p.FirstIsAverage,
		p.BarComponent, p.QuoteComponent, p.TradeComponent)
}

//nolint:funlen,cyclop
func newTripleExponentialMovingAverage(length int, alpha float64, firstIsAverage bool,
	bc entities.BarComponent, qc entities.QuoteComponent, tc entities.TradeComponent,
) (*TripleExponentialMovingAverage, error) {
	const (
		invalid = "invalid triple exponential moving average parameters"
		fmts    = "%s: %s"
		fmtw    = "%s: %w"
		fmtnl   = "tema(%d%s)"
		fmtna   = "tema(%d, %.8f%s)"
		minlen  = 2
		two     = 2.
		three   = 3
		epsilon = 0.00000001
	)

	var (
		mnemonic  string
		err       error
		barFunc   entities.BarFunc
		quoteFunc entities.QuoteFunc
		tradeFunc entities.TradeFunc
	)

	// Resolve defaults for component functions.
	// A zero value means "use default, don't show in mnemonic".
	if bc == 0 {
		bc = entities.DefaultBarComponent
	}

	if qc == 0 {
		qc = entities.DefaultQuoteComponent
	}

	if tc == 0 {
		tc = entities.DefaultTradeComponent
	}

	if math.IsNaN(alpha) {
		if length < minlen {
			return nil, fmt.Errorf(fmts, invalid, "length should be greater than 1")
		}

		alpha = two / float64(1+length)
		mnemonic = fmt.Sprintf(fmtnl, length, core.ComponentTripleMnemonic(bc, qc, tc))
	} else {
		if alpha < 0. || alpha > 1. {
			return nil, fmt.Errorf(fmts, invalid, "smoothing factor should be in range [0, 1]")
		}

		if alpha < epsilon {
			alpha = epsilon
		}

		length = int(math.Round(two/alpha)) - 1
		mnemonic = fmt.Sprintf(fmtna, length, alpha, core.ComponentTripleMnemonic(bc, qc, tc))
	}

	if barFunc, err = entities.BarComponentFunc(bc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	if quoteFunc, err = entities.QuoteComponentFunc(qc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	if tradeFunc, err = entities.TradeComponentFunc(tc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	desc := "Triple exponential moving average " + mnemonic

	tema := &TripleExponentialMovingAverage{
		smoothingFactor: alpha,
		length:          length,
		length2:         int(two)*length - 1,
		length3:         three*length - int(two),
		firstIsAverage:  firstIsAverage,
	}

	tema.LineIndicator = core.NewLineIndicator(mnemonic, desc, barFunc, quoteFunc, tradeFunc, tema.Update)

	return tema, nil
}

// IsPrimed indicates whether an indicator is primed.
func (s *TripleExponentialMovingAverage) IsPrimed() bool {
	s.mu.RLock()
	defer s.mu.RUnlock()

	return s.primed
}

// Metadata describes an output data of the indicator.
// It always has a single scalar output -- the calculated value of the triple exponential moving average.
func (s *TripleExponentialMovingAverage) Metadata() core.Metadata {
	return core.Metadata{
		Type:        core.TripleExponentialMovingAverage,
		Mnemonic:    s.LineIndicator.Mnemonic,
		Description: s.LineIndicator.Description,
		Outputs: []outputs.Metadata{
			{
				Kind:        int(TripleExponentialMovingAverageValue),
				Type:        outputs.ScalarType,
				Mnemonic:    s.LineIndicator.Mnemonic,
				Description: s.LineIndicator.Description,
			},
		},
	}
}

// Update updates the value of the indicator given the next sample.
func (s *TripleExponentialMovingAverage) Update(sample float64) float64 { //nolint:cyclop,funlen
	const three = 3.

	if math.IsNaN(sample) {
		return sample
	}

	s.mu.Lock()
	defer s.mu.Unlock()

	if s.primed {
		sf := s.smoothingFactor
		v1 := s.ema1
		v2 := s.ema2
		v3 := s.ema3
		v1 += (sample - v1) * sf
		v2 += (v1 - v2) * sf
		v3 += (v2 - v3) * sf
		s.ema1 = v1
		s.ema2 = v2
		s.ema3 = v3

		return three*(v1-v2) + v3
	}

	s.count++
	if s.firstIsAverage { //nolint:nestif
		if s.count == 1 {
			s.sum = sample
		} else if s.length >= s.count {
			s.sum += sample
			if s.length == s.count {
				s.ema1 = s.sum / float64(s.length)
				s.sum = s.ema1
			}
		} else if s.length2 >= s.count {
			s.ema1 += (sample - s.ema1) * s.smoothingFactor
			s.sum += s.ema1

			if s.length2 == s.count {
				s.ema2 = s.sum / float64(s.length)
				s.sum = s.ema2
			}
		} else { // if s.length3 >= s.count {
			s.ema1 += (sample - s.ema1) * s.smoothingFactor
			s.ema2 += (s.ema1 - s.ema2) * s.smoothingFactor
			s.sum += s.ema2

			if s.length3 == s.count {
				s.primed = true
				s.ema3 = s.sum / float64(s.length)

				return three*(s.ema1-s.ema2) + s.ema3
			}
		}
	} else { // Metastock
		if s.count == 1 {
			s.ema1 = sample
		} else if s.length >= s.count {
			s.ema1 += (sample - s.ema1) * s.smoothingFactor
			if s.length == s.count {
				s.ema2 = s.ema1
			}
		} else if s.length2 >= s.count {
			s.ema1 += (sample - s.ema1) * s.smoothingFactor
			s.ema2 += (s.ema1 - s.ema2) * s.smoothingFactor

			if s.length2 == s.count {
				s.ema3 = s.ema2
			}
		} else { // if s.length3 >= s.count {
			s.ema1 += (sample - s.ema1) * s.smoothingFactor
			s.ema2 += (s.ema1 - s.ema2) * s.smoothingFactor
			s.ema3 += (s.ema2 - s.ema3) * s.smoothingFactor

			if s.length3 == s.count {
				s.primed = true

				return three*(s.ema1-s.ema2) + s.ema3
			}
		}
	}

	return math.NaN()
}

File: ts/indicators/patrick-mulloy/triple-exponential-moving-average/triple-exponential-moving-average.ts
import { componentTripleMnemonic } from '../../core/component-triple-mnemonic';
import { IndicatorMetadata } from '../../core/indicator-metadata';
import { IndicatorType } from '../../core/indicator-type';
import { LineIndicator } from '../../core/line-indicator';
import { OutputType } from '../../core/outputs/output-type';
import { TripleExponentialMovingAverageOutput } from './triple-exponential-moving-average-output';
import { TripleExponentialMovingAverageLengthParams } from './triple-exponential-moving-average-params';
import { TripleExponentialMovingAverageSmoothingFactorParams } from './triple-exponential-moving-average-params';

const guardLength = (object: any): object is TripleExponentialMovingAverageLengthParams => 'length' in object;

/** Function to calculate mnemonic of a __TripleExponentialMovingAverage__ indicator. */
export const tripleExponentialMovingAverageMnemonic =
  (params: TripleExponentialMovingAverageLengthParams | TripleExponentialMovingAverageSmoothingFactorParams): string => {
  if (guardLength(params)) {
    const p = params as TripleExponentialMovingAverageLengthParams;
    return 'tema('.concat(Math.floor(p.length).toString(),
      componentTripleMnemonic(p.barComponent, p.quoteComponent, p.tradeComponent), ')');
  } else {
    const p = params as TripleExponentialMovingAverageSmoothingFactorParams;
    const length = Math.round(2 / p.smoothingFactor) - 1;
    return 'tema('.concat(length.toString(), ', ', p.smoothingFactor.toFixed(8),
      componentTripleMnemonic(p.barComponent, p.quoteComponent, p.tradeComponent), ')');
  }
};

// https://store.traders.com/-v12-c01-smoothi-pdf.html
// https://store.traders.com/-v12-c02-smoothi-pdf.html

/** __Triple Exponential Moving Average__ line indicator computes the triple exponential, or triple exponentially weighted, moving average (_TEMA_).
 *
 * The TEMA was developed by Patrick G. Mulloy and is described in two articles:
 *
 * ❶ Technical Analysis of Stocks &amp; Commodities v.12:1 (11-19), Smoothing Data With Faster Moving Averages.
 *
 * ❷ Technical Analysis of Stocks &amp; Commodities v.12:2 (72-80), Smoothing Data With Less Lag.
 *
 * The calculation is as follows:
 *
 * EMA¹ᵢ = EMA(Pᵢ) = αPᵢ + (1-α)EMA¹ᵢ₋₁ = EMA¹ᵢ₋₁ + α(Pᵢ - EMA¹ᵢ₋₁), 0 < α ≤ 1
 *
 * EMA²ᵢ = EMA(EMA¹ᵢ) = αEMA¹ᵢ + (1-α)EMA²ᵢ₋₁ = EMA²ᵢ₋₁ + α(EMA¹ᵢ - EMA²ᵢ₋₁), 0 < α ≤ 1
 *
 * EMA³ᵢ = EMA(EMA²ᵢ) = αEMA²ᵢ + (1-α)EMA³ᵢ₋₁ = EMA³ᵢ₋₁ + α(EMA²ᵢ - EMA³ᵢ₋₁), 0 < α ≤ 1
 *
 * TEMAᵢ = 3(EMA¹ᵢ - EMA²ᵢ) + EMA³ᵢ
 *
 * The very first EMA value (the seed for subsequent values) is calculated differently.
 * This implementation allows for two algorithms for this seed.
 *
 * ❶ Use a simple average of the first 'period'. This is the most widely documented approach.
 *
 * ❷ Use first sample value as a seed. This is used in Metastock.
 */
export class TripleExponentialMovingAverage extends LineIndicator {
  private readonly smoothingFactor: number;
  private readonly firstIsAverage: boolean;
  private readonly length: number;
  private readonly length2: number;
  private readonly length3: number;
  private sum = 0;
  private count = 0;
  private ema1 = 0;
  private ema2 = 0;
  private ema3 = 0;

  /**
   * Constructs an instance given a length in samples or a smoothing factor in (0, 1).
   **/
  public constructor(params: TripleExponentialMovingAverageLengthParams | TripleExponentialMovingAverageSmoothingFactorParams){
    super();
    let len;
    if (guardLength(params)) {
      const p = params as TripleExponentialMovingAverageLengthParams;
      len = Math.floor(p.length);
      if (len < 2) {
        throw new Error('length should be greater than 1');
      }

      this.length = len;
      this.smoothingFactor = 2 / (len + 1);
      this.firstIsAverage = p.firstIsAverage;

    } else {
      const p = params as TripleExponentialMovingAverageSmoothingFactorParams;
      if (p.smoothingFactor <= 0 || p.smoothingFactor >= 1) {
        throw new Error('smoothing factor should be in range (0, 1)');
      }

      this.smoothingFactor = p.smoothingFactor;
      this.length = Math.round(2 / this.smoothingFactor) - 1;
      this.firstIsAverage = false;
    }

    this.length2 = 2 * this.length - 1;
    this.length3 = 3 * this.length - 2;
    this.mnemonic = tripleExponentialMovingAverageMnemonic(params);
    this.description = 'Triple exponential moving average ' + this.mnemonic;
    this.barComponent = params.barComponent;
    this.quoteComponent = params.quoteComponent;
    this.tradeComponent = params.tradeComponent;
    this.primed = false;
  }

  /** Describes the output data of the indicator. */
  public metadata(): IndicatorMetadata {
    return {
      type: IndicatorType.TripleExponentialMovingAverage,
      mnemonic: this.mnemonic,
      description: this.description,
      outputs: [{
        kind: TripleExponentialMovingAverageOutput.TripleExponentialMovingAverageValue,
        type: OutputType.Scalar,
        mnemonic: this.mnemonic,
        description: this.description,
      }],
    };
  }

  /** Updates the value of the indicator given the next sample. */
  public update(sample: number): number {
    if (Number.isNaN(sample)) {
      return sample;
    }

    if (this.primed) {
      this.ema1 += (sample - this.ema1) * this.smoothingFactor;
      this.ema2 += (this.ema1 - this.ema2) * this.smoothingFactor;
      this.ema3 += (this.ema2 - this.ema3) * this.smoothingFactor;
      return 3 * (this.ema1 - this.ema2) + this.ema3;
    }

    // Not primed.
    ++this.count;
    if (this.firstIsAverage) { // First is the simple average.
      if (this.count === 1) {
        this.sum = sample;
      } else if (this.length >= this.count) {
        this.sum += sample;
        if (this.length === this.count) {
          this.ema1 = this.sum / this.length;
          this.sum = this.ema1;
        }
      } else if (this.length2 >= this.count) {
        this.ema1 += (sample - this.ema1) * this.smoothingFactor;
        this.sum += this.ema1;
        if (this.length2 === this.count) {
          this.ema2 = this.sum / this.length;
          this.sum = this.ema2;
        }
      } else { //if (this.length3 >= this.count) {
        this.ema1 += (sample - this.ema1) * this.smoothingFactor;
        this.ema2 += (this.ema1 - this.ema2) * this.smoothingFactor;
        this.sum += this.ema2;
        if (this.length3 === this.count) {
          this.primed = true;
          this.ema3 = this.sum / this.length;
          return 3 * (this.ema1 - this.ema2) + this.ema3;
        }
      }
    } else { // firstIsAverage is false, Metastock case.
      if (this.count === 1) {
        this.ema1 = sample;
      } else if (this.length >= this.count) {
        this.ema1 += (sample - this.ema1) * this.smoothingFactor;
        if (this.length === this.count) {
          this.ema2 = this.ema1;
        }
      } else if (this.length2 >= this.count) {
        this.ema1 += (sample - this.ema1) * this.smoothingFactor;
        this.ema2 += (this.ema1 - this.ema2) * this.smoothingFactor;
        if (this.length2 === this.count) {
          this.ema3 = this.ema2;
        }
      } else { //if (this.length3 >= this.count) {
        this.ema1 += (sample - this.ema1) * this.smoothingFactor;
        this.ema2 += (this.ema1 - this.ema2) * this.smoothingFactor;
        this.ema3 += (this.ema2 - this.ema3) * this.smoothingFactor;
        if (this.length3 === this.count) {
          this.primed = true;
          return 3 * (this.ema1 - this.ema2) + this.ema3;
        }
      }
    }

    return Number.NaN;
  }
}
